### PR TITLE
Add aligned_accessor proposal

### PR DESCRIPTION
**Pull request by mhoemmen**
I plan to finish this today.  It's still a draft, because I need to finish formatting.

Fixes https://github.com/ORNL/cpp-proposals-pub/issues/370

### CODE REVIEW

File: aligned_accessor/aligned_accessor.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+document: P1673
+date: today
+audience: LEWG
+author:
+  - name: Mark Hoemmen
+    email: <mhoemmen@nvidia.com>
+  - name: Damien Lebrun-Grandie
+    email: <lebrungrandt@ornl.gov>
+  - name: Nicolas Manual Morales
+    email: <???@sandia.gov>
```
Comment (dalg24) [inline aligned_accessor/aligned_accessor.md:13]:
```suggestion
    email: < nmmoral@sandia.gov>
```

File: aligned_accessor/aligned_accessor.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
+
+---
+title: "A free function linear algebra interface based on the BLAS"
+document: P1673
```
Comment (dalg24) [inline aligned_accessor/aligned_accessor.md:4]: Need to update this

File: aligned_accessor/aligned_accessor.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+toc: true
+---
+
+# Authors
+
+* Mark Hoemmen (mhoemmen@nvidia.com) (NVIDIA)
+
+* Damien Lebrun-Grandie (lebrungrandt@ornl.gov) (Oak Ridge National Laboratory)
+
+* Nicolas Manuel Morales (???@sandia.gov) (Sandia National Laboratories)
```
Comment (dalg24) [inline aligned_accessor/aligned_accessor.md:25]:
```suggestion
* Nicolas Manuel Morales (nmmoral@sandia.gov) (Sandia National Laboratories)
```

**Review by dalg24 — COMMENTED** [review]
I see you added an example and actual wording.  Did you change anything in the rest?

File: aligned_accessor/aligned_accessor.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  using reference = ElementType&;
+  using data_handle_type = ElementType*;
+
+  static constexpr size_t byte_alignment = the_byte_alignment;
+
+  constexpr aligned_accessor() noexcept = default;
+
+  template<class OtherElementType, size_t other_byte_alignment>
+    constexpr aligned_accessor(
+      aligned_accessor<OtherElementType, other_byte_alignment>) noexcept
```
Comment (crtrott) [inline aligned_accessor/aligned_accessor.md:232]:
```suggestion
      aligned_accessor<OtherElementType, other_byte_alignment>) noexcept;
```

Comment (crtrott) [reply aligned_accessor/aligned_accessor.md:232]: and then delete the next line. Inline provided functions don't have a full explanation later generally

File: aligned_accessor/aligned_accessor.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+// Interfaces that require 32-byte alignment,
+// because they want to do 8-wide SIMD of float.
+extern void vectorized_axpy(aligned_mdspan<32> y, float alpha, aligned_mdspan<32> x);
+extern float vectorized_norm(aligned_mdspan<32> y);
+
+// Interfaces that require 16-byte alignment,
+// because they want to do 4-wide SIMD of float.
+extern void fill_x(aligned_mdspan<32> x);
+extern void fill_y(aligned_mdspan<32> y);
```
Comment (dalg24) [inline aligned_accessor/aligned_accessor.md:147]: Should that read 16 instead of 32?

File: aligned_accessor/aligned_accessor.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+  constexpr operator default_accessor<element_type>() const {
+    return {};
+  }
+
+  constexpr reference access(data_handle_type p, size_t i) const noexcept;
+
+  constexpr typename offset_policy::data_handle_type
+    offset(data_handle_type p, size_t i) const noexcept;
+
+  constexpr static bool is_sufficiently_aligned(data_handle_type p) {
```
Comment (crtrott) [inline aligned_accessor/aligned_accessor.md:244]: remove the implementation here, and provide what ti does in the description.

**Review by crtrott — COMMENTED** [review]
Looks pretty good.

File: aligned_accessor/aligned_accessor.md (hunk 6 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md]: Done; thanks!

File: aligned_accessor/aligned_accessor.md (hunk 1 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md]: Done; thanks!

File: aligned_accessor/aligned_accessor.md (hunk 2 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md]: Fixed; thanks! I'll get the paper number later today.

File: aligned_accessor/aligned_accessor.md (hunk 3 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md]: Fixed; thanks!

**Comment by mhoemmen** [comment]
Implementation: https://godbolt.org/z/8W86nYG9h

File: aligned_accessor/aligned_accessor.md (hunk 5 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md]: Fixed; thanks!

File: aligned_accessor/aligned_accessor.md (hunk 4 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md]: Fixed; thanks!

File: aligned_accessor/aligned_accessor.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+template<class OtherElementType, size_t other_byte_alignment>
+  constexpr aligned_accessor(
+    aligned_accessor<OtherElementType, other_byte_alignment>) noexcept;
+```
+
+[1]{.pnum} *Constraints*:
+
+* [1.1]{.pnum} `is_convertible_v<OtherElementType(*)[], element_type(*)[]>` is `true`, and
+
+* [1.2]{.pnum} `gcd(other_byte_alignment, byte_alignment) == byte_alignment` is `true`.
```
Comment (dalg24) [inline aligned_accessor/aligned_accessor.md:268]: Effects: ...

File: aligned_accessor/aligned_accessor.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+[2]{.pnum} `aligned_accessor` meets the accessor policy requirements.
+
+[3]{.pnum} `ElementType`  is required to be a complete object type that is neither an abstract class type nor an array type.
+
+[4]{.pnum} Each specialization of `aligned_accessor` is a trivially copyable type that models `semiregular`.
+
+[5]{.pnum} $[0, n)$ is an accessible range for an object `p` of type `data_handle_type` and an object of type `aligned_accessor` if and only if $[$`p`, `p` + $n)$ is a valid range.
+
+## Members [mdspan.accessor.aligned.members]
+
```
Comment (dalg24) [inline aligned_accessor/aligned_accessor.md:257]: Default constructor is missing

File: aligned_accessor/aligned_accessor.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+  constexpr aligned_accessor(
+    aligned_accessor<OtherElementType, other_byte_alignment>) noexcept;
+```
+
+[1]{.pnum} *Constraints*:
+
+* [1.1]{.pnum} `is_convertible_v<OtherElementType(*)[], element_type(*)[]>` is `true`, and
+
+* [1.2]{.pnum} `gcd(other_byte_alignment, byte_alignment) == byte_alignment` is `true`.
+
```
Comment (dalg24) [inline aligned_accessor/aligned_accessor.md:269]: Conversion operator to default_accessor is also missing

File: aligned_accessor/aligned_accessor.md (hunk 8 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md:262]: Thanks for noticing this, though I think it's OK.  ` constexpr aligned_accessor() noexcept = default;`  is in the class overview.  Also, the wording is consistent with `default_accessor`, which doesn't have a separate entry for the defaulted default constructor: 
https://eel.is/c++draft/views.multidim#mdspan.accessor.default .

File: aligned_accessor/aligned_accessor.md (hunk 9 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md:274]: It's defined inline in the class overview, so we shouldn't need wording for it.

File: aligned_accessor/aligned_accessor.md (hunk 7 above)
Comment (mhoemmen) [reply aligned_accessor/aligned_accessor.md:273]: Thanks for pointing this out!  To be consistent with the wording for `default_accessor`, which doesn't have an *Effects* clause, I've just added here that the definition is `{}`.

File: aligned_accessor/aligned_accessor.md (hunk 8 above)
Comment (dalg24) [reply aligned_accessor/aligned_accessor.md:262]: You are right.  Sorry about the noise :)

**Review by dalg24 — APPROVED** [review]

File: aligned_accessor/aligned_accessor.md (hunk 7 above)
Comment (crtrott) [reply aligned_accessor/aligned_accessor.md:273]: Yeah I think that is good enough for now. LWG can tell us how to spell this.

**Review by crtrott — APPROVED** [review]

**Pull request merged**